lightning/src/util/fairrwlock.rs

```diff
@@ -0,0 +1,42 @@
+use std::sync::{TryLockResult, LockResult, RwLock, RwLockReadGuard, RwLockWriteGuard};
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+/// Rust libstd's RwLock does not provide any fairness guarantees (and, in fact, when used on
+/// Linux with pthreads under the hood, readers trivially and completely starve writers).
+/// Because we often hold read locks while doing message processing in multiple threads which
+/// can use significant CPU time, with write locks being time-sensitive but relatively small in
+/// CPU time, we can end up with starvation completely blocking incoming connections or pings,
+/// especially during initial graph sync.
+///
+/// Thus, we need to block readers when a writer is pending, which we do with a trivial RwLock
+/// wrapper here. Its not particularly optimized, but provides some reasonable fairness by
+/// blocking readers (by taking the write lock) if there are writers pending when we go to take
+/// a read lock.
+pub struct FairRwLock<T> {
+	lock: RwLock<T>,
+	waiting_writers: AtomicUsize,
+}
+
+impl<T> FairRwLock<T> {
+	pub fn new(t: T) -> Self {
+		Self { lock: RwLock::new(t), waiting_writers: AtomicUsize::new(0) }
+	}
+
+	pub fn write(&self) -> LockResult<RwLockWriteGuard<T>> {
+		self.waiting_writers.fetch_add(1, Ordering::AcqRel);
+		let res = self.lock.write();
+		self.waiting_writers.fetch_sub(1, Ordering::AcqRel);
+		res
+	}
+
+	pub fn try_write(&self) -> TryLockResult<RwLockWriteGuard<T>> {
+		self.lock.try_write()
+	}
+
+	pub fn read(&self) -> LockResult<RwLockReadGuard<T>> {
+		if self.waiting_writers.load(Ordering::Acquire) != 0 {
+			let _write_queue_lock = self.lock.write();
```

Comment:
I think a comment here would be nice explaining the idea of initiating a wait for a write lock before a new read lock is accessed.

Reply:
The current mechanism seems to achieve that whenever there are pending write locks, a new read lock cannot be added until those pending write locks complete. However, if there are already pending read locks, and a new write lock gets added, it will have to wait. Have you considered ways to jump the line, so to speak, and would that be a desirable property?

Reply:
Jumping the line would break the fairness property - if there's waiting writers, and we allow new readers to take the read lock, then those new readers may cause the waiting writer to wait even longer, which is the issue we're trying to solve to begin with.

Reply:
As for a comment, I'm not sure what to add to the existing one that's in the struct docs:

```
/// Thus, we need to block readers when a writer is pending, which we do with a trivial RwLock
/// wrapper here. Its not particularly optimized, but provides some reasonable fairness by
/// blocking readers (by taking the write lock) if there are writers pending when we go to take
/// a read lock.
```